pkg/importer/gcs-datasource.go

```diff
@@ -0,0 +1,200 @@
+		bucket, object, host = extractGcsBucketObjectAndHost(endpoint)
+		options = append(options, option.WithEndpoint(host))
+	}
+
+	// Creating GCS Client
+	client, err := getGcsClient(ctx, keyFile, options...)
+
+	if err != nil {
+		klog.Errorf("GCS Importer: Error creating GCS Client")
+		return nil, err
+	}
+
+	// Creating GCS Reader
+	gcsReader, err := newReaderFunc(ctx, client, bucket, object)
+	if err != nil {
+		klog.Errorf("GCS Importer: Error creating Reader")
+		return nil, err
+	}
+
+	return &GCSDataSource{
+		ep:        ep,
+		keyFile:   keyFile,
+		gcsReader: gcsReader,
+	}, nil
+
+}
+
+// Info is called to get initial information about the data.
+func (sd *GCSDataSource) Info() (ProcessingPhase, error) {
+	var err error
+	sd.readers, err = NewFormatReaders(sd.gcsReader, uint64(0))
+	if err != nil {
+		klog.Errorf("GCS Importer: Error creating readers: %v", err)
+		return ProcessingPhaseError, err
+	}
+	if !sd.readers.Convert {
+		// Downloading a raw file, we can write that directly to the target.
+		return ProcessingPhaseTransferDataFile, nil
+	}
+
+	return ProcessingPhaseTransferScratch, nil
+}
+
+// Transfer is called to transfer the data from the source to a temporary location.
+func (sd *GCSDataSource) Transfer(path string) (ProcessingPhase, error) {
+	klog.V(3).Infoln("GCS Importer: Transfer")
+
+	size, _ := util.GetAvailableSpace(path)
+
+	if size <= int64(0) {
+		//Path provided is invalid.
+		klog.V(3).Infoln("GCS Importer: Transfer Error: ", ErrInvalidPath)
+		return ProcessingPhaseError, ErrInvalidPath
+	}
+
+	file := filepath.Join(path, tempFile)
+	err := util.StreamDataToFile(sd.readers.TopReader(), file)
+
+	if err != nil {
+		klog.V(3).Infoln("GCS Importer: Transfer Error: ", err)
+		return ProcessingPhaseError, err
+	}
+	// If streaming succeeded, then parsing the file into URL will also succeed, no need to check error status
+	sd.url, _ = url.Parse(file)
+	return ProcessingPhaseConvert, nil
+}
+
+// TransferFile is called to transfer the data from the source to the passed in file.
+func (sd *GCSDataSource) TransferFile(fileName string) (ProcessingPhase, error) {
+	err := util.StreamDataToFile(sd.readers.TopReader(), fileName)
+	if err != nil {
+		return ProcessingPhaseError, err
+	}
+	return ProcessingPhaseResize, nil
+}
+
+// GetURL returns the url that the data processor can use when converting the data.
+func (sd *GCSDataSource) GetURL() *url.URL {
+	return sd.url
+}
+
+// Close closes any readers or other open resources.
+func (sd *GCSDataSource) Close() error {
+	var err error
+	if sd.readers != nil {
+		err = sd.readers.Close()
+	}
+	return err
+}
+
+// Create a Cloud Storage Client
+func getGcsClient(ctx context.Context, keyFile string, options ...option.ClientOption) (*storage.Client, error) {
+	klog.V(3).Infoln("GCS Importer: Creating Client")
+	if keyFile != "" {
+		klog.V(3).Infoln("GCS Importer: Creating Client: Authenticated")
+		return storage.NewClient(ctx, options...)
+	}
```

Comment:
are you missing an else here? Seems like you are always adding `option.WithoutAuthentication()` even if keyFile is not empty.

Reply:
Only append if keyfile== ""